File: backend-go/db/redis.go
package db

import (
	"context"
	"fmt"
	"time"

	"github.com/momentohq/client-sdk-go/auth"
	"github.com/momentohq/client-sdk-go/config"
	"github.com/momentohq/client-sdk-go/config/logger"
	"github.com/momentohq/client-sdk-go/momento"
	"github.com/momentohq/client-sdk-go/responses"
)

const CACHE_NAME = "spreadsheet_cache"

type Redis struct {
	Client *momento.CacheClient
}

func NewRedis(ctx context.Context) *Redis {
	return &Redis{
		Client: initializeRedis(ctx),
	}
}

func initializeRedis(ctx context.Context) *momento.CacheClient {
	credentialProvider, err := auth.NewEnvMomentoTokenProvider("MOMENTO_AUTH_TOKEN")
	if err != nil {
		return nil
	}

	client, err := momento.NewCacheClient(
		config.LaptopLatestWithLogger(logger.NewNoopMomentoLoggerFactory()).WithClientTimeout(15*time.Second),
		credentialProvider,
		600*time.Second,
	)
	if err != nil {
		return nil
	}

	return &client
}

func (redis *Redis) Set(ctx context.Context, key, value string) error {
	_, err := (*redis.Client).Set(ctx, &momento.SetRequest{
		CacheName: CACHE_NAME,
		Key:       momento.String(key),
		Value:     momento.String(value),
		Ttl:       7 * 3600 * time.Second,
	})

	if err != nil {
		return err
	}

	return nil
}

func (redis *Redis) Get(ctx context.Context, key string) (string, error) {
	res, err := (*redis.Client).Get(ctx, &momento.GetRequest{
		CacheName: CACHE_NAME,
		Key:       momento.String(key),
	})
	if err != nil {
		return "", err
	}

	switch r := res.(type) {
	case *responses.GetHit:
		{
			return r.ValueString(), nil
		}
	case *responses.GetMiss:
		{
			return "", nil
		}
	}

	return "", nil
}

func (redis *Redis) Delete(ctx context.Context, key string) error {
	_, err := (*redis.Client).Delete(ctx, &momento.DeleteRequest{
		CacheName: CACHE_NAME,
		Key:       momento.String(key),
	})
	if err != nil {
		return err
	}

	return nil
}

func (redis *Redis) AuthKey(access_token string) string {
	return fmt.Sprintf("USER#%s", access_token)
}

func (redis *Redis) SpreadSheetKey(spreadsheet_id string) string {
	return fmt.Sprintf("SPREADSHEET#%s", spreadsheet_id)
}

File: backend-rust/src/db/redis.rs
use std::time::Duration;

use crate::config::SPREAD_SHEET_CACHE;

pub struct Redis {
    client: momento::SimpleCacheClient,
}

impl Redis {
    pub fn new() -> Self {
        // Initialize Momento
        let credential_provider =
            momento::CredentialProviderBuilder::from_environment_variable("MOMENTO_AUTH_TOKEN".to_string())
                .build()
                .expect("env var MOMENTO_AUTH_TOKEN must be set to your auth token");
        let item_default_ttl_seconds = 600;
        let cache_client = momento::SimpleCacheClientBuilder::new(
            credential_provider,
            Duration::from_secs(item_default_ttl_seconds),
        )
        .unwrap()
        .build();

        return Self {
            client: cache_client,
        };
    }

    pub async fn set(&mut self, key: &str, value: &str) -> momento::MomentoResult<()> {
        match self
            .client
            .set(
                SPREAD_SHEET_CACHE,
                key,
                value,
                Duration::from_secs(7 * 3600),
            )
            .await
        {
            Ok(_) => Ok(()),
            Err(err) => Err(err),
        }
    }

    pub async fn get(&mut self, key: &str) -> momento::MomentoResult<String> {
        match self.client.get(SPREAD_SHEET_CACHE, key).await {
            Ok(res) => match res {
                momento::response::Get::Hit { value } => Ok(value.try_into().expect("string")),
                momento::response::Get::Miss => Ok("".to_string()),
            },
            Err(err) => Err(err),
        }
    }

    pub fn auth_key(access_token: String) -> String {
        return format!("USER#{}", access_token);
    }

    pub fn spread_sheet_key(spreadsheet_id: String) -> String {
        return format!("SPREADSHEET#{}", spreadsheet_id);
    }
}
